taichi/rhi/impl_support.h

```diff
@@ -89,5 +91,39 @@ struct BidirMap {
   }
 };
 
+// A synchronized list of objects that is pointer stable & reuse objects
+// It does not mark objects as used, and it does not free objects (destructor is
+// not called)
+template <class T>
+class SyncedPtrStableObjectList {
```

Comment:
so to my understanding the the contained objects should be explicitly constructed/destructed?

Reply:
That's a good point. It seems like we should add a check that `T` is a POD-ish type, which doesn't have fancy ctor/dtor. (Dtor is especially important here, IMO. Right now the implementation would require the caller to call `~T()`)

Reply:
Fixed it now, the the object list now handles ctor/dtor & uses RAII.

Reply:
nit: I suggest that we add some unit test on this (https://github.com/taichi-dev/taichi/tree/master/tests/cpp), given that this is a very low-level infra, and has certain behaviors that is not trivial (manually calling `~T()`, placement new, ...)